mne/rank.py

```diff
@@ -4,23 +4,50 @@
 #
 # License: BSD (3-clause)
 
-# import operator
+import operator
 
 import numpy as np
 from scipy import linalg
 
-# from .io.pick import pick_info, _picks_by_type, _pick_data_channels
-from .io.pick import _picks_by_type
 from .defaults import _handle_default
-# from .io.proj import make_projector
-from .utils import logger
-from .utils import _compute_row_norms
-# from .proj import setup_proj
-from mne.utils import _apply_scaling_cov, _undo_scaling_cov
-from mne.utils import _apply_scaling_array, _undo_scaling_array
+from .io.meas_info import _simplify_info
+from .io.pick import _picks_by_type, pick_info, pick_channels_cov
+from .io.proj import make_projector
+from .utils import (logger, _compute_row_norms,
+                    _apply_scaling_cov, _undo_scaling_cov,
+                    _scaled_array,
+                    warn, _check_rank, verbose)
+
+
+# XXX : incorporate, add tests showing utility
+def _estimate_rank_by_ratio(X, ratio=1.0e4):
+    X = np.asarray(X)
+    assert X.ndim == 2
+    # Compute the singular values
+    sing = linalg.svd(X, compute_uv=False)
+    # If the largest singular value is zero, the rank is zero
+    if sing[0] < np.finfo(X.dtype).eps:
+        return 0
+    # Scale and check for zero singular values within the numerical
+    # precision and remove them
+    sing_scaled = sing / sing[0]
+    n_nonzero = np.count_nonzero(sing_scaled > np.finfo(X.dtype).eps)
+    sing = sing[:n_nonzero]
+    # Search for abrupt changes in the singular value spectra by
+    # computing the ratios of consecutive singular values
+    sing_shifted = np.hstack([sing[0], sing[:-1]])
+    sing_ratio = sing_shifted / sing
+    # Check if any/some of the ratios exceed the threshold
+    ex = np.where(sing_ratio > ratio)[0]
+    if len(ex) == 0:
+        return n_nonzero
+    else:
+        return ex[0]
```

Comment:
shall we remove this code for now?

Reply:
I'll just comment it out for now since it's on the imminent to-do list

Reply:
ok with an XXX